Add explanation on Python docstring format (#139)
lab-cosmo/equistore: CONTRIBUTING.rst

    """
    ...
    return result

Guidelines for writing Python doc strings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Use Python typing in the function arguments, indicate return types.

* Start the description after each ``:param:`` or ``:return:`` in a new line and add an
  empty line between the parameter and return block.

* Emphasize function and class parameters with a single backtick i.e ```param``` and
  general variables should be double backticked . i.e. ````my_variable````

* If you include any maths, make the string a
  `raw string`_ by prefixing with ``r``, e.g.,

  .. code-block:: python

    r"""Some math like :math:`\nu^2 / \rho` with backslashes."""

  Otherwise the ``\n`` and ``\r`` will be rendered as ASCII escape sequences that break
  lines without you noticing it or you will get either one of the following two
  errors message

  1. `Explicit markup ends without a blank line; unexpected unindent`
  2. `Inline interpreted text or phrase reference start-string without end string`

* The examples are tested with `doctest`_. Therefore, please make sure that they are
  complete and functioning (with all required imports). As given in the example above 

>   use the ``>>>`` syntax for inputs (followed by ``...`` for multiline inputs) and no

Reviewer: I'd move the `...` example in the example doc string above to here, as it has nothing to do with the example `func()`

* Use Python typing in the function arguments, indicate return types.

* Start the description after each ``:param:`` or ``:return:`` in a new line and add an
  empty line between the parameter and return block.

* Emphasize function and class parameters with a single backtick i.e ```param``` and
  general variables should be double backticked . i.e. ````my_variable````

* If you include any maths, make the string a
  `raw string`_ by prefixing with ``r``, e.g.,

  .. code-block:: python

    r"""Some math like :math:`\nu^2 / \rho` with backslashes."""

  Otherwise the ``\n`` and ``\r`` will be rendered as ASCII escape sequences that break
  lines without you noticing it or you will get either one of the following two
  errors message

  1. `Explicit markup ends without a blank line; unexpected unindent`
  2. `Inline interpreted text or phrase reference start-string without end string`

* The examples are tested with `doctest`_. Therefore, please make sure that they are
  complete and functioning (with all required imports).
  Use the ``>>>`` syntax for inputs (followed by ``...`` for multiline inputs) and no
  indentation for outputs for the examples.

  .. code-block:: python

>       """"
>       >>> a = np.array(
>       ...    [1, 2, 3, 4]
>       ... )
>       """"

Reviewer: ```suggestion
      """
      >>> a = np.array(
      ...    [1, 2, 3, 4]
      ... )
      """
```

    def func(value_1: float, value_2: int) -> float:
        r"""A one line summary sentence of the function.

        Extensive multi-line summary of what is going in. Use single 
        backticks for parameters of the function like `width` and two ticks for
        values ``67``. You can link to classes :py:class:`equistore.Labels`. This
        also works for other classes and functions like :py:obj:`True`.

        Inline Math is also possible with :math:`\mathsf{R}`. Or as a math block.

        .. math::

            \mathbf{x}' = \mathsf{R}\mathbf{x}

        :param value_1:
            The first parameter of the function, a :py:class:`float`.
        :param value_2:
            The second parameter of the function, an :py:class:`int`.

        :returns result:
            The result of the calculation, a :py:class:`float`.

        :raises TypeError:
            If `value_1` is not a :py:class:`float` or `value_2` is not a :py:class:`int`.
        :raises ValueError:
            If `value_1` is not greater than zero.

        Examples
        --------

        This is an example

        >>> from equistore import func
        >>> func(1, 1)
        42

>       """

Reviewer: Sorry these lines, 177, 178, 179, still need to be indented.

The documentation of equistore is written in reStructuredText (rst) and uses the
`sphinx`_ documentation generator. In order to modify the documentation, first
create a local version of the code on your machine as described above. Then, you
can build the documentation with:

.. code-block:: bash

    tox -e docs

You can visualize the local documentation with your favorite browser (here
Mozilla Firefox is used)

.. code-block:: bash

    firefox docs/build/html/index.html

.. _`sphinx` : https://www.sphinx-doc.org/en/master/

Python doc strings
~~~~~~~~~~~~~~~~~~

Our docstring format follows the `sphinx format`_ and a typical function doc string
looks like the following.

> .. code-block:: rst

Reviewer: Do you think it would be useful to include the actual function here too, so the context is complete?

For instance:
```py
def func(value_1: float, value_2: int) -> float:
    """
    <insert docstring here>
    """
    ...
    return result
```

Reviewer: On this note we could also highlight somewhere in this contributing documentation that contributors should use Python typing in the function arguments, indicate return types etc.

>     ONE LINE summary sentence.

Reviewer: ```suggestion
    A one line summary sentence of the function.
```

    Extensive multi-line summary of what is going in. Use single 
    backticks for parameters of the function like `width` and two ticks for
    values ``67``. You can link to classes :py:class:`equistore.Labels`. This
    also works for other classes and functions like :py:obj:`True`.

    Inline Math is also possible with :math:`\mathsf{R}`. Or as a math block.

    .. math::

        \mathbf{x}' = \mathsf{R}\mathbf{x}

>     :param value:

Reviewer: ```suggestion
    :param value_1:
```

>         The first value of the function.

Reviewer: ```suggestion
        The first parameter of the function, a :py:class:`float`.
```

Reviewer: Also, how/where should we indicate types in the docstring? Like this ^?

>     :param value:

Reviewer: ```suggestion
    :param value_2:
```

>         second parameter

Reviewer: ```suggestion
        The second parameter of the function, an :py:class:`int`.
```

>     :returns prev:

Reviewer: ```suggestion
    :returns result:
```

>         the result of the calculation

Reviewer: ```suggestion
        The result of the calculation, a :py:class:`float`.
```

    :raises ValueError:
        if the value is not a float

    Examples
    --------

    This is an example

    >>> from equistore import func
    >>> func(1, 1)
    42

> Some Notes on writing Python doc strings

Reviewer: ```suggestion
Guidelines for writing Python doc strings
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Start the description after each ``:param:`` or ``:return:`` in a new line and add an
  empty line between the parameter and return block.

* Emphasize function and class parameters with a single backtick i.e ```param``` and
  general variables should be double backticked . i.e. ````my_variable````

* If you include any maths, make the string a
  `raw string`_ by prefixing with ``r``, e.g.,

  .. code-block:: python

    r"""Some math like :math:`\nu^2 / \rho` with backslashes."""

  Otherwise the ``\n`` and ``\r`` will be rendered as ASCII escape sequences that break
  lines without you noticing it or you will get either one of the following two
  errors message

  1. `Explicit markup ends without a blank line; unexpected unindent`
  2. `Inline interpreted text or phrase reference start-string without end string`

> * The examples are testest with `doctest`_. Therefore, please use a syntax using ``>>>``

Reviewer: ```suggestion
* The examples are tested with `doctest`_. Therefore, please make sure that they are complete and functioning (with all required imports), and use the ``>>>`` syntax. 
```

>   for inputs (followed by ``...`` for multiline inputs) and no indentation for outputs

Reviewer: Could you include an example of how to use ``...``?
